src/analysis/calculus/inverse.lean

```diff
@@ -0,0 +1,582 @@
+
+variable {f}
+
+@[simp] lemma to_local_homeomorph_to_fun : (hf.to_local_homeomorph f).to_fun = f := rfl
+
+lemma mem_to_local_homeomorph_source : a ∈ (hf.to_local_homeomorph f).source :=
+  (classical.some_spec hf.approximates_deriv_on_open_nhds).fst.1
+
+variables (f f' a)
+
+/-- Given a function `f` with an invertible derivative, returns a function that is locally inverse
+to `f`. -/
+def local_inverse : F → E := (hf.to_local_homeomorph f).inv_fun
+
+variables {f f' a}
+
+lemma eventually_left_inverse : ∀ᶠ x in 𝓝 a, hf.local_inverse f f' a (f x) = x :=
+(hf.to_local_homeomorph f).eventually_left_inverse hf.mem_to_local_homeomorph_source
+
+lemma eventually_right_inverse : ∀ᶠ y in 𝓝 (f a), f (hf.local_inverse f f' a y) = y :=
+(hf.to_local_homeomorph f).eventually_right_inverse' hf.mem_to_local_homeomorph_source
+
+/-- If `f` has an invertible derivative `f'` at `a` in the sense of strict differentiability `(hf)`,
+then the inverse function `hf.local_inverse f` has derivative `f'.symm` at `f a`. -/
+theorem to_local_inverse :
+  has_strict_fderiv_at (hf.local_inverse f f' a) (f'.symm : F →L[𝕜] E) (f a) :=
+begin
+  dunfold local_inverse,
+  set e := hf.to_local_homeomorph f,
+  have ha := hf.mem_to_local_homeomorph_source,
+  refine ((e.prod e).is_o_congr (mk_mem_prod (e.map_source ha) (e.map_source ha))).2 _,
+  simp only [local_homeomorph.prod_to_local_equiv, continuous_linear_equiv.coe_apply,
+    local_equiv.prod_inv_fun, local_equiv.prod_to_fun, e.left_inv ha, (∘)],
+  suffices : is_o (λ (p : E × E), f'.symm (f' (p.fst - p.snd) - (e.to_fun p.fst - e.to_fun p.snd)))
+    (λ (p : E × E), e.to_fun p.fst - e.to_fun p.snd) (𝓝 (a, a)),
+  { refine this.congr' _ (eventually_of_forall _ $ λ _, rfl),
+    filter_upwards [continuous_fst.tendsto (a, a) (e.eventually_left_inverse ha),
+      continuous_snd.tendsto (a, a) (e.eventually_left_inverse ha)],
+    simp only [mem_set_of_eq, mem_preimage],
+    intros,
+    simp only [*, continuous_linear_equiv.map_sub, f'.symm_apply_apply] },
+  suffices : is_o (λ (p : E × E), (f' (p.fst - p.snd) - (f p.fst - f p.snd)))
+    (λ (p : E × E), f p.fst - f p.snd) (𝓝 (a, a)),
+  from (f'.symm.to_continuous_linear_map.is_O_comp _ _).trans_is_o this,
+  refine (hf.trans_is_O _).symm,
+  rcases hf.approximates_deriv_on_open_nhds with ⟨s, ⟨has, hs⟩, H⟩,
+  have := H.antilipschitz (f'.subsingleton_or_nnnorm_symm_pos.imp id $
+    λ hf', nnreal.half_lt_self $ ne_of_gt $ nnreal.inv_pos.2 $ hf'),
+  exact is_O.of_bound this.K
+    (eventually.mono (mem_nhds_sets (is_open_prod hs hs) (mk_mem_prod has has)) $
+      λ p hp, by { simp only [← dist_eq_norm], exact this.le_mul_dist ⟨p.1, hp.1⟩ ⟨p.2, hp.2⟩ })
+end
+
+/-- If `f : E → F` has an invertible derivative `f'` at `a` in the sense of strict differentiability
+and `g (f x) = x` in a neighborhood of `a`, then `g` has derivative `f'.symm` at `f a`. -/
+theorem to_local_inverse' {g : F → E} (hg : ∀ᶠ x in 𝓝 a, g (f x) = x) :
```

Comment:
The logical order between `to_local_inverse` and `to_local_inverse'` feels strange. Couldn't you prove directly `to_local_inverse'` (with weaker assumptions -- I guess you don't need completeness) and then deduce `to_local_inverse`?

Reply:
I remember that I had some difficulties with this. BTW, do you remember an example proving that `complete_space E` is required for the inverse function theorem?

Reply:
Here is an example. 

Fix two countable dense subsets of the real line `A` and `B`. One can construct a `C^1` diffeo of `ℝ` mapping `A` to `B`, as follows. Enumerate the points of `A` as `x_n`. Define `f` as a series `\sum f_n`, where `f_0 = id`,  `||f_n||_{C^1} \leq 10^{-n}`, and `f_n` is supported on a small enough neighborhood of `x_n` that does not contain `x_0`, ..., `x_{n-1}`, to adjust the value of `f (x_n)` without changing the value of `f` at the previous points. One can also ensure `f' (A) ⊆ B` with the same construction (and one could make sure that `f(A)` is exactly `B`, and that `f` is `C^∞`, but I won't need it).

Take `A = ℚ` and `B = ℚ - {1/n, n > 0}`. Then, seen as a map from `ℚ` to `ℚ`, `f` is stricly differentiable (as this is the case from `ℝ` to `ℝ`), but it is not locally onto around `0`.

Reply:
In this example `f` has a left inverse but this left inverse is not continuous, so `to_local_inverse'` needs `[complete_space E]`.

Reply:
Yes, I forgot to say that this example convinced me that one indeed needs `[complete_space E]` for this statement, so your logical flow of arguments is fine. Probably it wouldn't be needed for the same theorem with a right inverse instead?

Reply:
Rewriting...

Reply:
Done. I also noticed that this part works for `has_fderiv_at`.